cpp/src/io/shp/polygon_shapefile_reader.cpp

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cuspatial/error.hpp>
+#include <ogrsf_frmts.h>
+
+#include <cudf/types.hpp>
+
+#include <memory>
+#include <string>
+#include <tuple>
+
+namespace {
+
+cudf::size_type read_ring(OGRLinearRing const& ring,
+                          std::vector<double>& xs,
+                          std::vector<double>& ys)
+{
+    cudf::size_type num_vertices = ring.getNumPoints();
+
+    // append points in reverse order
```

Comment:
Why in reverse order?

Reply:
Good question. That was the original implementation, I believe. I’ll see if I can find out _why_, though.

Reply:
@zhangjianting may know the answer.